kong/clustering/services/sync/hooks.lua

```diff
@@ -0,0 +1,179 @@
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local hooks = require("kong.hooks")
+local EMPTY = require("kong.tools.table").EMPTY
+
+
+local ipairs = ipairs
+local ngx_null = ngx.null
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+
+
+local DEFAULT_PAGE_SIZE = 512
+
+
+function _M.new(strategy)
+  local self = {
+    strategy = strategy,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+local function get_all_nodes_with_sync_cap()
+  local res, err = kong.db.clustering_data_planes:page(DEFAULT_PAGE_SIZE)
+  if err then
+    return nil, "unable to query DB " .. err
+  end
+
+  if not res then
+    return EMPTY
+  end
+
+  local ret = {}
+  local ret_n = 0
+
+  for _, row in ipairs(res) do
+    for _, c in ipairs(row.rpc_capabilities) do
+      if c == "kong.sync.v2" then
+        ret_n = ret_n + 1
+        ret[ret_n] = row.id
+        break
+      end
+    end
+  end
+
+  return ret
+end
+
+
+function _M:notify_all_nodes()
```

Comment:
Is this behavior specific to lua-cp? It seems Koko might not reliably reach all data planes. If I understand correctly, there's a sync every 30 seconds where each dataplane fetches new deltas from the control plane. Isn't 30 seconds quite a long time? What's the approach for Koko -> DP delta notifications?

Reply:
Yes, 30 is a worst case fallback. Koko will be able to send the notify most of the time, and that is good enough for our use case.

Also, the notify call is idempotent, and Koko will not attempt to de-dup for that reasons. Since we have more than one RPC connection per DP node, the delivery rate will be good enough that 30 second delay will unlikely really be relied.

Reply:
That's correct! 30s is the worst case, the CP will use the `notify_new_version` when a new version is created (with loose delivery guarantees but that will effectively be more than 99% of the time that it is delivered).

Reply:
And also our intention is to trigger an immediate pull upon RPC reconnect, but this is not yet implemented in the tech preview.

Reply:
KAG-5569